File: bun/lib/tetradic.ts
export function makeTetradicNumber(index: number): number {
  return (index * (index + 1) * (index + 2)) / 6;
}

const ONE_THIRD = 1 / 3;

// This function is surprisingly fast
function getTetradicNumberIndex(y: number): number {
  const halfD = -3 * y;

  const discriminant = (halfD ** 2) + Math.pow(-ONE_THIRD, 3);
  const discriminantSqrt = Math.sqrt(discriminant);

  const u = Math.cbrt(-halfD + discriminantSqrt);
  const v = Math.cbrt(-halfD - discriminantSqrt);
  return u + v - 1;
}

export function getLowestTetradicNumberIndex(number: number): number {
  return Math.round(getTetradicNumberIndex(number));
}

File: cpp/tetradic.cpp
#include <math.h>

int makeTetradicNumber(int index) {
  return (index * (index + 1) * (index + 2)) / 6;
}

const double ONE_THIRD = 1.0 / 3.0;
const double MINUS_ONE_THIRD_CUBE = pow(-ONE_THIRD, 3);

// This function is surprisingly fast
double getTetradicNumberIndex(double y) {
  double halfD = -3 * y;

  double discriminant = (halfD * halfD) + MINUS_ONE_THIRD_CUBE;
  double discriminantSqrt = sqrt(discriminant);

  double u = pow(-halfD + discriminantSqrt, ONE_THIRD);
  double v = pow(-halfD - discriminantSqrt, ONE_THIRD);
  return u + v - 1;
}

int getLowestTetradicNumberIndex(double number) {
  return round(getTetradicNumberIndex(number));
}

